### PR TITLE
[TRTLLM-16404][fix] thread LoRA params through MoE models

### PR DESCRIPTION
**Pull request by achartier**
## Dev Engineer Review

- Threads optional `lora_params` through AFMoE, GLM-4 MoE, MiniMax M2/M3, and Step3p7 decoder, attention, MLP, and expert paths.
- Preserves MoE routing, shared-expert behavior, CUDA-graph slot handling, and caller compatibility.
- Applies Step3p7 routed scaling once and selects clamp-capable paths when required.
- Updates type annotations and docstrings without material scope changes.
- No material regression identified.

## QA Engineer Review

- Updates one integration helper and two unit-test files.
- Adds propagation tests for AFMoE, GLM-4, MiniMax M2/M3, and Step3p7.
- Covers shared and routed experts, dense MLPs, Step3p7 clamping and scaling, and attention QKV/output projections.
- The integration test is listed in `tests/integration/test_lists/test-db/l0_h100.yml` and `tests/integration/test_lists/qa/llm_function_core.txt`. Related A100 waivers remain in `tests/integration/test_lists/waives.txt`.
- Targeted tests passed 19 tests. A broader run passed 78 tests. Nine baseline or environment failures also reproduced on `origin/main`.
- Coverage verdict: **sufficient**.

### Per-File QA Perspective

- `tensorrt_llm/_torch/models/modeling_afmoe.py`: Verify `lora_params` propagation through attention, shared experts, dense MLPs, and routed experts.
- `tensorrt_llm/_torch/models/modeling_glm.py`: Verify propagation through GLM-4 MoE, decoder, MTP, attention, shared experts, and dense MLP paths.
- `tensorrt_llm/_torch/models/modeling_minimaxm2.py`: Verify decoder and model forwarding through attention and routed-expert paths.
- `tensorrt_llm/_torch/models/modeling_minimaxm3.py`: Verify propagation through shared and dense MLP paths.
- `tensorrt_llm/_torch/models/modeling_step3p7.py`: Verify attention LoRA, clamped projections, expert selection, routed scaling, and fused expert forwarding.
- `tensorrt_llm/_torch/peft/lora/cuda_graph_lora_params.py`: Verify type and docstring changes do not alter MoE slot-cache behavior.
- `tests/integration/defs/llmapi/test_llm_api_pytorch_moe_lora.py`: Updates helper annotations and documentation. Related integration tests remain listed in CI and manual-QA lists.
- `tests/unittest/_torch/peft/test_moe_lora_cuda_graph_params.py`: Updates helper annotations only. Runtime behavior is unchanged.
- `tests/unittest/_torch/peft/test_moe_lora_model_path.py`: Adds propagation and Step3p7 regression coverage. These unit tests have no separate integration-list entries.

## Description

Follow up on unresolved review comments from #18527 that still apply on `main`.

Several newer MoE model wrappers accepted request-level LoRA state at the model entry point but did not forward `lora_params` through their decoder layers into the routed-expert implementation. As a result, routed-expert LoRA adapters could be silently skipped for AFMoE, GLM-4 MoE, MiniMax M2/M3, and Step3p7.

The routed-scaling refactor also corrects a latent double-scaling issue for non-LoRA CUTLASS execution by keeping routing weights unscaled and applying `routed_scaling_factor` exactly once at the common MoE output.

This change:

- threads `lora_params` through those model-specific decoder and MoE paths;
- adds regression coverage that verifies every affected wrapper passes the exact LoRA state to its expert module;
- retains the CUDA-graph MoE slot behavior added by #18527 and addresses remaining typing/docstring review comments in that code and its tests.

## Test Coverage

- Targeted MoE LoRA tests on H100: 19 passed.
- Broader related unit-test run on H100: 78 passed, 9 failed; all 9 failures reproduced unchanged on `origin/main` and are environment/baseline failures unrelated to this patch.
- Pre-commit checks: passed.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/models/modeling_afmoe.py [hunk 1]
```diff
@@ -368,14 +371,15 @@ def forward(
...
+            lora_params=lora_params,
             **kwargs,
         )
         hidden_states = self.post_attention_layernorm(hidden_states)
 
         hidden_states, residual = self.pre_mlp_layernorm(hidden_states, residual)
 
         if self.moe_enabled:
-            hidden_states = self.mlp(hidden_states, attn_metadata)
+            hidden_states = self.mlp(hidden_states, attn_metadata, lora_params=lora_params)
```
[the remark this answers is not in the document]
Comment (achartier) [reply tensorrt_llm/_torch/models/modeling_afmoe.py:383]: Addressed in f2cb423f49. Added a parameterized decoder-layer regression that runs the real forward path for AFMoE, MiniMax M2, MiniMax M3, GLM4, and Step3p7 and asserts the request LoRA object reaches the MoE boundary unchanged. The focused installed-wheel suite passes (25 tests).

File: tensorrt_llm/_torch/models/modeling_step3p7.py [hunk 2]
```diff
@@ -884,6 +884,7 @@ def forward(
         hidden_states: torch.Tensor,
         attn_metadata: AttentionMetadata,
         all_reduce_params: Optional[AllReduceParams] = None,
+        lora_params: Optional[dict] = None,
```
[the remark this answers is not in the document]
Comment (achartier) [reply tensorrt_llm/_torch/models/modeling_step3p7.py:940]: Addressed in f2cb423f49. Step3p7 now records whether routed-expert LoRA is configured and bypasses the Python clamp fallback in that case; the fused MoE path receives both the LoRA params and an explicit SwigluActivation clamp. Added a regression with both clamp flags and MoE LoRA enabled. The focused installed-wheel suite passes (25 tests).

**Comment by achartier** [comment]
Addressed the new Step3p7 routed-scaling finding in 23be393fb1706f99683fb5733bb9670aebb67788.

Step3p7 separated routing now keeps normalized weights unscaled, and moe_router_scaling_factor is applied exactly once in the common post-expert path for both CUTLASS and the Python clamp fallback.

Added a numerical regression with moe_router_scaling_factor=3.0 through real Step3p7MoE construction. The old behavior would produce 9.0; the fixed path produces the intended 3.0.

Validation on the Computelab H100 using the installed full-wheel environment: 26 passed across the focused MoE-LoRA suites.

**Comment by achartier** [comment]
Addressed the outside-diff Step3p7 attention-LoRA finding in bc39a61a735702259812ad521be52f62800d1791.

Step3p7Attention now merges request LoRA into QKV with LoraLayer.forward_with_base, passes has_lora to the attention backend, and forwards lora_params plus layer_idx to the output projection while retaining the model-specific Q/K normalization and head gate.

Added a numerical regression with configured attn_qkv and attn_dense targets. Base projections produce zero; the adapter paths contribute a non-zero expected output, and the test also verifies backend and output-projection forwarding.

Validation in the Computelab H100 full-wheel environment: 27 passed across the focused MoE-LoRA suites. Pre-commit passed.

**Comment by achartier** [comment]
@brnguyen2 this is a follow-up on the remaining items from https://github.com/NVIDIA/TensorRT-LLM/pull/18527

File: tensorrt_llm/_torch/models/modeling_step3p7.py [hunk 3]
```diff
@@ -669,13 +685,47 @@ def __init__(
...
+                lambda: self.gate_up_proj(hidden_states),
+                (self.splitted_gate_up_lora, self.fused_gate_up_lora),
+                hidden_states,
+                lora_params,
+                self.layer_idx,
+            )
+        else:
+            gate_up = self.gate_up_proj(hidden_states)
+        gate, up = gate_up.chunk(2, dim=-1)
+        gate = torch.nn.functional.silu(gate.clamp(max=self.swiglu_limit))
```
Comment (BowenFu) [inline tensorrt_llm/_torch/models/modeling_step3p7.py]: The no-LoRA path changes from clamping the SiLU output to clamping its input. For a gate value above the limit these are different, and `_python_clamped_moe_forward` below still uses SiLU-then-clamp, so this makes the LoRA plumbing alter existing Step3p7 numerics and leaves the dense and routed clamped paths inconsistent.

```suggestion
        gate = torch.nn.functional.silu(gate).clamp(max=self.swiglu_limit)
```

I think preserving the existing activation order is required for this PR.

**Review by lori-ren — APPROVED** [review]

File: tensorrt_llm/_torch/models/modeling_step3p7.py (hunk 3 above)
Comment (achartier) [reply tensorrt_llm/_torch/models/modeling_step3p7.py]: Thanks for the catch. The agent changed the clamp order to be consistent with the CUTLASS SwiGLU implementation, but after checking the official step3p7 implementation, the CUTLASS implementation is the wrong one. I will address that in a separate PR.

File: tensorrt_llm/_torch/models/modeling_step3p7.py [hunk 4]
```diff
@@ -739,6 +791,7 @@ def __init__(
             model_config=model_config,
             layer_idx=layer_idx,
             weight_loading_mode=MoEWeightLoadingMode.VANILLA,
+            activation=SwigluActivation(clamp=self._routed_swiglu_limit),
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/models/modeling_step3p7.py:794]: The clamp the fused path applies here is not the same operation as `_python_clamped_moe_forward` / `ClampedGatedMLP`. The thop op promotes `Swiglu` with a `swiglu_limit` to the `SwigluBias` adaptor (`cpp/tensorrt_llm/thop/moeOp.cpp:563-570`), and `SwigluBiasAdaptor` clamps the gate **before** the sigmoid: `min(g,L)·σ(min(g,L))` (`moe_kernels.cuh`). The Python path and the HF reference clamp **after** SiLU: `silu(g).clamp(max=L)`. The two agree for gate pre-activations ≤ L but diverge above it (bounded by `L·(1−σ(L))` per element — small at L≈7, material at small limits). Since `_moe_lora_enabled` is a static config check, merely configuring MoE-LoRA targets shifts base-model numerics on clamp layers even for requests carrying no adapter. Please either verify this is acceptable against the HF reference for real Step3p7 limits, or note the approximation in the comment above — the current wording ("joins the common output scaling") reads as if the paths were numerically equivalent.

File: tests/unittest/_torch/peft/test_moe_lora_model_path.py [hunk 5]
```diff
@@ -70,6 +83,446 @@ def test_mixtral_moe_forward_passes_lora_params_to_routed_experts():
     )
 
 
+@pytest.mark.parametrize(
+    "moe_cls",
+    [AfmoeMoE, MiniMaxM2MoE, MiniMaxM3MoE],
+)
+def test_model_moe_forward_passes_lora_params_to_routed_experts(
+    moe_cls: type[torch.nn.Module],
```
Comment (brnguyen2) [inline tests/unittest/_torch/peft/test_moe_lora_model_path.py:91]: All the new coverage here (and in test_modeling_step3p7.py) builds `SimpleNamespace` fake-selves with mocked experts, so it pins the Python call contract but can't catch anything at or below the fused op — e.g. the clamp-order semantics of `SwigluActivation(clamp=...)` on the CUTLASS path, or a wrong LoRA module-type wiring in the real constructed modules. Since this PR is what makes routed-expert LoRA reachable for these five models, consider adding at least one of them to the end-to-end routed-expert LoRA integration test (test_llm_api_pytorch_moe_lora.py only got annotation updates) as a follow-up.

**Review by brnguyen2 — APPROVED** [review]
Approving — the comments below are optional touch-ups, not blockers.

The routed-scaling refactor checks out end to end: routing weights are now unscaled everywhere, the factor is applied exactly once at the MoE output, and TRTLLMGen's MiniMaxM2 metadata branch passes `routed_scaling_factor=None` — this also fixes a latent double-scale on any backend that consumes `routing_method.apply()` (CUTLASS, exactly the backend MoE LoRA forces). Worth a sentence in the description, since it changes non-LoRA numerics on CUTLASS too.

Two things worth addressing (inline): the clamp-order difference between the fused CUTLASS path and the Python reference on clamp-active Step3p7 layers, and the lack of any end-to-end LoRA coverage for the five newly wired models — every new test mocks the expert module, so a kernel-level divergence is invisible to this suite.

Given this fixes silently-dropped LoRA adapters across five models, a tracking NVBug/JIRA in the title instead of `[None]` would help trace it later.

File: tests/unittest/_torch/peft/test_moe_lora_model_path.py (hunk 5 above)
Comment (achartier) [reply tests/unittest/_torch/peft/test_moe_lora_model_path.py:91]: Added as a follow-up item in TRTLLM-16404

File: tensorrt_llm/_torch/models/modeling_step3p7.py (hunk 4 above)
Comment (achartier) [reply tensorrt_llm/_torch/models/modeling_step3p7.py:794]: Added as a follow-up item in TRTLLM-16404

**Review by BowenFu — APPROVED** [review]
